### PR TITLE
Soft 98 front power distribution gpio

### PR DESCRIPTION
**Pull request by copperium**

### CODE REVIEW

File: projects/front_power_distribution/src/front_power_distribution_gpio.c [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+typedef enum {
+  OUTPUT_GPIO_PIN_DRIVER_DISPLAY = 0,
+  OUTPUT_GPIO_PIN_STEERING,
+  OUTPUT_GPIO_PIN_CENTRE_CONSOLE,
+  OUTPUT_GPIO_PIN_HEADLIGHTS,
+  OUTPUT_GPIO_PIN_SIGNAL_LEFT,
+  OUTPUT_GPIO_PIN_SIGNAL_RIGHT,
+  NUM_OUTPUT_GPIO_PINS,
+} OutputGpioPin;
```
Comment (ArshanKhanifar) [inline projects/front_power_distribution/src/front_power_distribution_gpio.c]: Change it to `FrontPowerDistributionGpioOutput`

File: projects/front_power_distribution/inc/front_power_distribution_gpio.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+// Receive GPIO events and set the corresponding GPIO pin states.
+// Requires GPIO to be initialized.
+
+#include "event_queue.h"
+#include "gpio.h"
+
+void front_power_distribution_gpio_init(void);
+
+// Call in main event-processing loop.
```
Comment (ArshanKhanifar) [inline projects/front_power_distribution/inc/front_power_distribution_gpio.h]: this comment isn't necessary... it's implied that `*process_event` functions are called in `main`. All of our codebase shows that.

File: projects/front_power_distribution/src/front_power_distribution_gpio.c [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  if (id == FRONT_POWER_DISTRIBUTION_GPIO_EVENT_SIGNAL_HAZARD) {
+    // special case: maps to both left and right signals
+    gpio_set_state(&s_output_gpio_pins[OUTPUT_SIGNAL_LEFT], new_state);
+    gpio_set_state(&s_output_gpio_pins[OUTPUT_SIGNAL_RIGHT], new_state);
+    return STATUS_CODE_OK;
+  } else if (FRONT_POWER_DISTRIBUTION_GPIO_EVENT_DRIVER_DISPLAY <= id &&
+             id < NUM_FRONT_POWER_DISTRIBUTION_GPIO_EVENTS) {
+    gpio_set_state(&s_output_gpio_pins[s_events_to_gpio_outputs[e->id]], new_state);
+    return STATUS_CODE_OK;
+  } else {
```
Comment (ArshanKhanifar) [inline projects/front_power_distribution/src/front_power_distribution_gpio.c]: Move this check to the beginning of the function. Usually it's a good idea to keep all the argument checking at the top of the function, for consistency :)

File: projects/front_power_distribution/test/test_front_power_distribution_gpio.c [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#include "front_power_distribution_events.h"
+#include "front_power_distribution_gpio.h"
+#include "log.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+static GpioAddress s_pin_addresses[] = {
```
Comment (ArshanKhanifar) [inline projects/front_power_distribution/test/test_front_power_distribution_gpio.c]: it's not a bad idea to expose a `GpioAddress * test_provide_gpio_addresses()`to your `front_power_distribution_gpio` module, that simply returns `s_output_gpio_pins`. That way you don't have to redefine everything here again.

Comment (copperium) [reply projects/front_power_distribution/test/test_front_power_distribution_gpio.c]: `s_output_gpio_pins` is indexed by `FrontPowerDistributionGpioOutput`s, not `FrontPowerDistributionGpioEvent`s, so I'd have to expose `s_events_to_gpio_outputs` in a similar function too. Should I do that?

**Pull request merged**